src/framework/hooks.js

```diff
@@ -0,0 +1,71 @@
+import { isFunction } from '../utils';
```

Comment:
If you think that students will read this code, than I would suggest to add a lot of comments to this file.

Even to me it took some time to understand what is going on here. Students 100% will fail.